Add additional features

DarkmatterVale opened the pull request:
This PR is for the GCI task located at https://codein.withgoogle.com/dashboard/task-instances/4892859836661760/?sp-page=1. The features updated in this PR includes:
- Updated tip selection process. Each tip is now only used once every "rotation" (all tips in the tip file)
- Added Icons for each category
- Added hyperlinks to the icons
- Fixed a bug which does not allow the application to find tips

DarkmatterVale committed:
Fix issues when no file is found on system

DarkmatterVale committed:
Add Icon artwork

DarkmatterVale committed:
Added hyperlink functionality

DarkmatterVale committed:
Add icons

DarkmatterVale committed:
Fix incorrect icon name

DarkmatterVale committed:
Update tip selection process

jessicah commented on Tipster/Tipster.cpp:
```diff
@@ -92,7 +126,7 @@ Tipster::MouseDown(BPoint pt)
 	GetMouse(&temp, &buttons);
 	
-	if (Bounds().Contains(temp)) {
+	if (Bounds().Contains(temp)) {		
```
Erroneous trailing whitespace

jessicah commented on Tipster/Tipster.cpp:
```diff
@@ -22,20 +24,25 @@
+	UPDATE_ICON = 'upin'
 };
 
+#define BROWSER_MIME_WEBPOSITIVE "application/x-vnd.Haiku-WebPositive"\
```
Erroneous trailing slash

jessicah commented on Tipster/Tipster.cpp:
```diff
@@ -104,8 +138,39 @@ Tipster::MouseDown(BPoint pt)
+	fUpdatedList.Remove(0);
+	
+	BString link = fUpdatedList.StringAt(2);
+	//int32 linkoffset = TextLength();
```
Don't commit commented text. That is what your commit history is for.

jessicah commented on Tipster/Tipster.cpp:
```diff
@@ -134,6 +199,24 @@ Tipster::GetTipsFile()
 }
 
 
+/*Tipster::tLink*
```
Again, don't commit commented code.

jessicah commented on Tipster/Tipster.h:
```diff
@@ -21,20 +21,46 @@ class Tipster : public BTextView
 		void UpdateTip(void);
 		void LoadTips(entry_ref ref);
-
+		
```
Trailing whitespace again

DarkmatterVale committed:
Updated README.md picture

jessicah commented on Tipster/Tipster.h:
```diff
@@ -21,20 +21,46 @@ class Tipster : public BTextView
+		
+		BMessenger* messenger;
+		
+		static text_run_array linkStyle;
```
Why does the link style need to be static? Do you have multiple instances of the Tipster class? And itt's modified in a non-static member function (AttachedToWindow())...

DarkmatterVale replied:
This is the way PulkoMandy used the linkStyle in fRiSS

humdingerb commented on Tipster/MainWindow.h:
```diff
@@ -22,9 +24,14 @@ class MainWindow : public BWindow
+		BButton* icon;
+		BString* url;
+		BBitmap* icon_bitmap;
 };
```
The above variables should be named fIconButton, fURL and fIconBitmap or something alike.

humdingerb commented on Tipster/MainWindow.cpp:
```diff
@@ -87,6 +107,31 @@ for Haiku");
+				BPath path("artwork");
+				path.Append(artwork);
+				
+				icon_bitmap = BTranslationUtils::GetBitmap(path.Path());
```
As Scott commented already on the GCI task, you should use ressources from an rdef instead.

humdingerb commented on Tipster/MainWindow.cpp:
```diff
@@ -55,9 +61,23 @@ MainWindow::BuildLayout(void)
+	icon->SetFlat(true);
+	
+	static const float spacing = be_control_look->DefaultItemSpacing() / 2;
+	fMainSplitView = 
```
No need for a BSplitView.

DarkmatterVale replied:
How else do I show the Icon to the left of the view?

humdingerb commented on Tipster/MainWindow.cpp:
```diff
@@ -87,6 +107,31 @@ for Haiku");
+			fTipsterView->OpenURL(url);
+			
+			break;
+		}
```
Maybe remove a few unnecessary empty lines from this switch block.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -85,6 +101,22 @@ Tipster::MessageReceived(BMessage* msg)
+	char *argv[2];
+	BString app;
+
+	app = BROWSER_MIME_WEBPOSITIVE;
```
Instead of hard-wiring Web+, please use Launch() with the HTTP URL MIME type (application/x-vnd.Be.URL.http) to find the user's preferred app.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -85,6 +101,22 @@ Tipster::MessageReceived(BMessage* msg)
+	argv[0] = (char*)url->String();
+	argv[1] = 0;
+
+	status_t status = be_roster->Launch( app.String(), 1, argv );
```
No spaces after/before those brackets.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,6 +56,17 @@ Tipster::AttachedToWindow()
+	linkStyle.count = 1;
+	linkStyle.runs[0].offset = 0;
+	linkStyle.runs[0].font = linkfont;
+	linkStyle.runs[0].color = make_color(0,0,255);
```
Spaces after the commas.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -152,14 +207,21 @@ Tipster::LoadTips(entry_ref ref)
+	}
+	else if (category == "Preferences") {
+		return "App_Workspaces";
+	}
```
No curly brackets for single line if/else-blocks.

humdingerb commented on Tipster/Tipster.h:
```diff
@@ -15,26 +15,52 @@ class Tipster : public BTextView
 {
 	public:
 		Tipster(void);
```
No "void"s, a simple "()" should do.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -152,14 +207,21 @@ Tipster::LoadTips(entry_ref ref)
+const char * 
+Tipster::GetArtworkTitle(BString category)
+{
+	if (category == "GUI") {
```
category "Application" is missing

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -104,8 +136,31 @@ Tipster::MouseDown(BPoint pt)
 void
 Tipster::UpdateTip()
```
Currently tipster-tips.txt needs to start with a "%", else there's some mess-up that'll end up showing the URL of the first tip. Since the "fortunes" format doesn't have a "%" at the start, you should try to not need it for Tipster. We may load other fortunes-style files in the future.

DarkmatterVale replied:
This has been fixed @humdingerb

humdingerb commented:
The correctly line-broken tip is only displayed after resizing the window. To see that, just launch Tipster and, if needed, click thru the tips for one that's longer than fits in the default window.
Maybe some Invalidate() missing somewhere? I dunno...

DarkmatterVale committed:
Update artwork to load from rdef file

DarkmatterVale committed:
Implement proper tip reading, miscellaneous updates

DarkmatterVale commented:
@humdingerb This has been fixed in the latest commit.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
+	linkStyle.runs[0].font = linkfont;
+	linkStyle.runs[0].color = make_color(0, 0, 255);
+
+	messenger = new BMessenger(this->Parent());
```
Should be fMessenger (also in .h)

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
+	if (file.InitCheck() != B_OK)
+		return;
+
+	BString fTip;
```
Only member variables would be named fTip.  Maybe use just "tip" instead.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
+	file.Read(buf, size);
+	fTip.UnlockBuffer(size);
+	
+	BStringList fUpdatedList;
```
Same here. Maybe "updatedList" instead.

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
+	BString link = fUpdatedList.StringAt(2);
+	
+	Insert(fUpdatedList.StringAt(1));
+	Insert("\n\n<-- Click on the icon to open the URL");
```
The whole AddBeginningTip() thing seems to be a bit convoluted. Why not always pick the first tip in the tips-file as the start tip?

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
+}
+
+
+/*void
```
Remove Tipster::OnResize() completely then.

humdingerb commented on Tipster/Tipster.h:
```diff
@@ -14,27 +14,55 @@
+		tLink* GetLinkAt(BPoint point);
+		const char* GetArtworkTitle(BString category);
+
+		BList links;
```
"links" appears not to be used.

humdingerb commented on Tipster/Tipster.h:
```diff
@@ -14,27 +14,55 @@
+		BMessenger* messenger;
+
+		text_run_array linkStyle;
 };
```
Add tabs to have all the variable names are aligned in one column.

humdingerb commented on Tipster/Tipster.h:
```diff
@@ -14,27 +14,55 @@
+		BString* fIntroductionTip;
+
 		bigtime_t fTime;
 		BMessageRunner *fRunner;
```
BMessageRunner\* fRunner;

humdingerb commented on Tipster/Tipster.cpp:
```diff
@@ -51,12 +56,72 @@ Tipster::AttachedToWindow()
 {
 	BMessage message(M_CHECK_TIME);
 	fRunner = new BMessageRunner(this, &message, 1000000);
+	
```
That linkstyle thing isn't needed anymore, right?

DarkmatterVale committed:
Code cleanup

DarkmatterVale committed:
Add proper introduction tip

DarkmatterVale committed:
Don't allow user to Vertically resize

DarkmatterVale committed:
Fix bug with lone URL

DarkmatterVale commented:
@humdingerb I think I have cleaned all the code up.

I refactored the AddBeginningTip() method to make it cleaner and reuse other functions within Tipster.
